[UI] Implemented category and description attributes for codes

hyperbx opened the pull request:
This PR extends the Codesv2 framework by adding two new keywords to the language for category and description attributes.

These keywords include;
- `in` for categories.
- `does` for descriptions.

These keywords can be used in any order, but `does` is recommended to be placed at the end.

Example usage;
```
Patch "Name" in "Category" by "Author" does "Description"
```

The description attribute also allows multiline descriptions;
```
Patch "Name" in "Category" by "Author" does
/*
Description
*/
```

The syntax used for multiline descriptions is to maintain compatibility with older versions of Hedge Mod Manager and newer code files, as old versions will read the multiline description as a multiline comment when compiling it as C#.

This PR includes categories and descriptions for all pre-existing codes, these have been tested on an older version of Hedge Mod Manager and do not cause any issues in terms of backwards compatibility.

UI changes;
- A new column has been added to the codes list for Category.
![image](https://user-images.githubusercontent.com/34012267/204139286-a879a9d8-f3da-416e-acfc-fca7c8a82461.png)
- Clicking a column in the codes list will sort the codes in either ascending or descending order based on the contents of it.
![image](https://user-images.githubusercontent.com/34012267/204139470-dbe4e05d-4de9-432d-8141-0c03d3573fb6.png)
- Codes no longer have drag and drop support in the codes list.
- Columns can no longer be rearranged in the codes list.
- Double-clicking a code will open an information window containing the name, author and description attributes.
![image](https://user-images.githubusercontent.com/34012267/204139335-2ea09efe-6408-4f85-a0af-f738aa69bb60.png)
- Hovering over a code will display a tooltip to hint about double-clicking.
![image](https://user-images.githubusercontent.com/34012267/204139328-79efb391-957f-42b1-bee8-15365c27a0c9.png)

Future goals;
- Localisation support for code names, categories and descriptions (Codesv3).

hyperbx committed:
[Codes] Implemented keywords for Category and Description

hyperbx committed:
[UI] Category and Description UI for Codes

hyperbx committed:
[Codes] Update all codes with applicable fields

hyperbx committed:
[Sonic Origins] Use latest codes as base

oops

RagdollClash commented on HedgeModManager/Resources/Codesv2/SonicGenerations.hmm:
```diff
@@ -1,27 +1,29 @@
 Write<uint>(GetPointer(0x1E5E2F0, 0x5B8), 0x3E7)
 
-Patch "Disable Ring Cap" by "Darío"
+Patch "Disable Ring Cap" in "Cheats" by "Darío" does "Allows the ring counter to exceed 999 rings."
```
Wouldn't it be better for this to be under Gameplay?

hyperbx committed:
[Sonic Generations] Move Disable Ring Cap to Gameplay

hyperbx commented:
Currently reworking how the codes list is sorted, will update when it's done.

hyperbx committed:
[UI] Improved codes list column sorting

hyperbx commented:
Columns in the codes list will now sort by Category by default, with a new `CodesSortingColumnIndex` key added to the registry config so it remembers what your last sort was.

Codes now sort with the enabled codes first, then disabled codes last.

![image](https://user-images.githubusercontent.com/34012267/204181837-0badaf5f-38ba-4b44-9669-208edf0dfff5.png)

hyperbx committed:
[Codes] Added "Fixes" category

hyperbx committed:
[UI] Implemented code description on hover

HyperBE32 commented:
Descriptions for codes are now displayed at the bottom of the codes list upon hovering over any that have a description. This can be disabled in the Settings tab.

![aK1f5cZG1G](https://user-images.githubusercontent.com/34012267/204196568-042a50b1-2223-4cce-946d-82230a582cfc.gif)

hyperbx committed:
[UI] Improved code description box

If the mouse cursor intersects the lower description box, it'll now use a box at the top of the list instead. This fixes the issue with it flickering if the mouse cursor was intersecting the last visible item in the list with a description.

hyperbx committed:
[UI] Disable column re-ordering for Mods and Codes

Neither of these lists save the column order anywhere, so this feature is useless.

hyperbx committed:
MainWindow.xaml: remove unnecessary using for System.Reflection

hyperbx committed:
[UI] Use fixed bounds for code description

hyperbx committed:
[UI] Allow user to resize code description

hyperbx committed:
Merge remote-tracking branch 'upstream/rewrite' into rewrite

hyperbx committed:
Merge remote-tracking branch 'upstream/rewrite' into rewrite

hyperbx committed:
Merge remote-tracking branch 'upstream/rewrite' into rewrite

hyperbx committed:
Merge remote-tracking branch 'upstream/rewrite' into rewrite

hyperbx committed:
[UI] Clear code description on refresh

hyperbx committed:
[Codes] Move Author property to match metadata order (useless commit, don't look)

hyperbx committed:
[Codes] See last commit, same story

hyperbx committed:
[Languages] Remove unused localised strings for previous feature

hyperbx committed:
Merge remote-tracking branch 'upstream/rewrite' into rewrite

Sajidur78 reviewed (approved):
LGTM
